### PR TITLE
created DoubleSolenoidSubsystem

### PR DESCRIPTION
**Pull request by Spikes-2212**

### CODE REVIEW

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
+package com.spikes2212.command;
+
+import com.spikes2212.dashboard.Namespace;
+import com.spikes2212.dashboard.RootNamespace;
+import edu.wpi.first.networktables.NetworkTable;
+import edu.wpi.first.wpilibj.DoubleSolenoid;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+
+/**
+ * A class that represents a double solenoid.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:10]: 1. It represents a subsystem, not the solenoid itself.
2. B-I\a2

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * @author Camellia Lami
+ * @see DashboardedSubsystem
+ */
+public class DoubleSolenoidSubsystem extends DashboardedSubsystem {
+
+    /**
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:17]: Given a descriptive enough class documentation, this is unnecessary.

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public DoubleSolenoidSubsystem(String namespaceName, DoubleSolenoid doubleSolenoid) {
+        this(new RootNamespace(namespaceName), doubleSolenoid);
+    }
+
+    /**
+     * Opens the double solenoid.
+     *
+     * @return An {@link InstantCommand} that opens the solenoid.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:34]: B-I\b7

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+     */
+    public InstantCommand openSolenoid() {
+        return new InstantCommand(() -> doubleSolenoid.set(DoubleSolenoid.Value.kReverse));
+    }
+
+    /**
+     * Closes the double solenoid.
+     *
+     * @return An {@link InstantCommand} that closes the solenoid.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:43]: B-I\b7

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+ * @see DashboardedSubsystem
+ */
+public class DoubleSolenoidSubsystem extends DashboardedSubsystem {
+
+    /**
+     * Creates a {@link DoubleSolenoid}.
+     */
+    private final DoubleSolenoid doubleSolenoid;
+
+    public DoubleSolenoidSubsystem(Namespace namespace, DoubleSolenoid doubleSolenoid) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:22]: add inverted

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        super(namespace);
+        this.doubleSolenoid = doubleSolenoid;
+    }
+
+    public DoubleSolenoidSubsystem(String namespaceName, DoubleSolenoid doubleSolenoid) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:27]: add inverted

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        this(new RootNamespace(namespaceName), doubleSolenoid);
+    }
+
+    /**
+     * Opens the double solenoid.
+     *
+     * @return An {@link InstantCommand} that opens the solenoid.
+     */
+    public InstantCommand openSolenoid() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:36]: consider inverted

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        return new InstantCommand(() -> doubleSolenoid.set(DoubleSolenoid.Value.kReverse));
+    }
+
+    /**
+     * Closes the double solenoid.
+     *
+     * @return An {@link InstantCommand} that closes the solenoid.
+     */
+    public InstantCommand closeSolenoid() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:45]: consider inverted

**Review by Yoel-Perman-Brilliant — CHANGES_REQUESTED** [review]

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+import com.spikes2212.dashboard.Namespace;
+import com.spikes2212.dashboard.RootNamespace;
+import edu.wpi.first.networktables.NetworkTable;
+import edu.wpi.first.wpilibj.DoubleSolenoid;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import edu.wpi.first.wpilibj2.command.Subsystem;
+
+/**
+ * A double solenoid {@link Subsystem}
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:11]: Take inspiration from `SmartMotorControllerGenericSubsystem`

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * @author Camellia Lami
+ * @see DashboardedSubsystem
+ */
+public class DoubleSolenoidSubsystem extends DashboardedSubsystem {
+
+    private final DoubleSolenoid doubleSolenoid;
+    private final boolean isInverted;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:19]: change to `inverted` and add documentation for it

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+    public DoubleSolenoidSubsystem(Namespace namespace, DoubleSolenoid doubleSolenoid, boolean isInverted) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:21]: ditto

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        super(namespace);
+        this.doubleSolenoid = doubleSolenoid;
+        this.isInverted = isInverted;
+    }
+
+    public DoubleSolenoidSubsystem(String namespaceName, DoubleSolenoid doubleSolenoid, boolean isInverted) {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:27]: ditto

**Review by Yoel-Perman-Brilliant — CHANGES_REQUESTED** [review]

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+ * A {@link Subsystem} that controls an applicable double solenoid.
+ *
+ * @author Camellia Lami
+ * @see DashboardedSubsystem
+ */
+public class DoubleSolenoidSubsystem extends DashboardedSubsystem {
+
+    private final DoubleSolenoid doubleSolenoid;
+    /**
+     * Checks if the solenoid is inverted.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:20]: i think the description is a little unnecessary, not sure

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Opens the double solenoid.
+     *
+     * @return an {@link InstantCommand} that opens the solenoid
+     */
+    public InstantCommand openSolenoid() {
+        if (inverted) {
+            return new InstantCommand(() -> doubleSolenoid.set(DoubleSolenoid.Value.kForward));
+        }
+        else return new InstantCommand(() -> doubleSolenoid.set(DoubleSolenoid.Value.kReverse));
```
Comment (Microsh1t) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:40]: don't use 'else', just use return.

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Closes the double solenoid.
+     *
+     * @return an {@link InstantCommand} that closes the solenoid
+     */
+    public InstantCommand closeSolenoid() {
+        if (inverted) {
+            return new InstantCommand(() -> doubleSolenoid.set(DoubleSolenoid.Value.kReverse));
+        }
+        else return new InstantCommand(() -> doubleSolenoid.set(DoubleSolenoid.Value.kForward));
```
Comment (Microsh1t) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:52]: don't use 'else', just use return.

File: src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    public DoubleSolenoidSubsystem(String namespaceName, DoubleSolenoid doubleSolenoid, boolean inverted) {
+        this(new RootNamespace(namespaceName), doubleSolenoid, inverted);
+    }
+
+    /**
+     * Opens the double solenoid.
+     *
+     * @return an {@link InstantCommand} that opens the solenoid
```
Comment (Microsh1t) [inline src/main/java/com/spikes2212/command/DoubleSolenoidSubsystem.java:34]: return **a** link

**Review by Microsh1t — CHANGES_REQUESTED** [review]

**Review by Yoel-Perman-Brilliant — APPROVED** [review]

**Pull request merged**